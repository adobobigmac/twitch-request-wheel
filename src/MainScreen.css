#main-screen {
    display: flex;
}
#main-screen.game-select .left-column {
    width: 45vw;
}
#main-screen.player-select .left-column {
    width: 90vw;
}

#main-screen .left-column h2 {
    margin-bottom: 0;
}
#main-screen .left-column h4 {
    color: yellow;
    font-size: 20px;
    font-weight:  400;
    margin-bottom: 12px;
    margin-top: 6px;
}
#main-screen.player-select .subheading-game,
#main-screen.game-select .subheading-player {
    display: none;
}

.left-column .left-column-body {
    display: flex;
    align-items: flex-start;
    height: 100%;
}
.left-column .left-column-inner-body {
    flex-grow: 2;
    margin-left: 15px;
}

#main-screen .right-column {
    /* width: 50vw; */
    text-transform: capitalize;
}

.wheel-wrapper {
    font-size: 16px;
    overflow: hidden;
    width: 600px;
}
.wheel-wrapper canvas {
    cursor: pointer;
}

#main-screen .logout {
    position: absolute;
    top:  0;
    right:  0;
    background-color: darkcyan;
    border-radius: 5px;
    margin-top:  0;
    padding-bottom: 5px;
    padding-top: 5px;
    color: #fff;
}



#main-screen .overlay {
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: rgba(0,0,0,0.5);
}
#main-screen .confetti-wrapper {
    position: absolute;
    right: 50%;
    left: 50%;
    bottom: 50;
}
#main-screen .fade-in-out {
    opacity: 0;
    animation: 3.5s linear 0s main-modal-fade;
    animation-fill-mode: forwards;
}
#main-screen .modal {
    position: fixed;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    background: rgba(0,0,0,0.5);
    padding: 2vh 4vw;
    color: #fff;
}
@keyframes main-modal-fade {
    0%,100% { opacity: 0 }
    5%,95% { opacity: 1 }
}
