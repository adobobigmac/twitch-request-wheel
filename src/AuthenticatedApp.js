
import React, { Component } from 'react';
import MainScreen from './MainScreen';
import JackboxGameList from './JackboxGameList';
import { Redirect, withRouter } from "react-router-dom";
import queryString from 'query-string'
const fetch = require('node-fetch');

class AuthenticatedApp extends Component {
  constructor(){
    super();
    this.state = {
      username: localStorage.getItem('__username'),
      access_token: localStorage.getItem('__access_token'),
      failed_login: false
    }
    this.getAuth = this.getAuth.bind(this);
    this.logOut = this.logOut.bind(this);
  }
  componentDidMount() {
      if (!this.state.access_token) {
          return this.getAuth();
      }
      fetch('https://api.twitch.tv/helix/users', {
        headers: {
          'Client-ID': process.env.REACT_APP_TWITCH_CLIENT_ID,
          Authorization: `Bearer ${this.state.access_token}`
        }
      })
      .then(r => r.json())
      .then(userInfo => {
        //console.log(userInfo); //login [aka lowercase username?], display_name, profile_image_url, description

        localStorage.setItem('__username', userInfo.data[0].login);
        return fetch(`https://api.twitch.tv/helix/moderation/moderators?broadcaster_id=${userInfo.data[0].id}`, {
          headers: {
            'Client-ID': process.env.REACT_APP_TWITCH_CLIENT_ID,
            Authorization: `Bearer ${this.state.access_token}`
          }
        })
        .then(r => r.json())
        .then(modInfo => {
          const modList = modInfo.data.map((modObj) => modObj.user_name.toLowerCase())
          this.setState((state) => {
            return {
              ...state,
              username: userInfo.data[0].login,
              modList
            };
          })
        })
      })
      .catch(e => this.getAuth)
  }

  logOut() {
      localStorage.removeItem('__username');
      localStorage.removeItem('__access_token');
      fetch('https://id.twitch.tv/oauth2/revoke?' + new URLSearchParams({
        client_id: process.env.REACT_APP_TWITCH_CLIENT_ID,
        token: this.state.access_token,
        redirect_uri: process.env.REACT_APP_REDIRECT_URI_NOENCODE
      }), {
        method: 'POST',
        headers: {
          Accept: 'application/vnd.twitchtv.v5+json'
        }
      })
      .then(() => {
          window.location.reload();
      });
  }

  async getAuth(e) {
      if (e) {
          console.error(e);
      }
      localStorage.removeItem('__username');
      localStorage.removeItem('__access_token');

    const queryParams = queryString.parse(this.props.location.search);
    await fetch('https://id.twitch.tv/oauth2/token?' + new URLSearchParams({
      grant_type: 'authorization_code',
      code: queryParams.code,
      client_id: process.env.REACT_APP_TWITCH_CLIENT_ID,
      client_secret: process.env.REACT_APP_TWITCH_CLIENT_SECRET,
      redirect_uri: process.env.REACT_APP_REDIRECT_URI_NOENCODE
    }), {
      method: 'POST',
      headers: {
        Accept: 'application/vnd.twitchtv.v5+json'
      }
    }).then(r => r.json())
    .then((oauth) => {
      //console.log(oauth); // access_token, refresh_token, expires_in, scope ['...']
      if(!oauth.access_token) {
        this.setState((state) => {
          return {
            ...state,
            failed_login: true
          };
        });
        return;
      }

      localStorage.setItem('__access_token', oauth.access_token);
      this.setState((state) => {
        return {
          ...state,
          access_token: oauth.access_token
        };
      });

      fetch('https://api.twitch.tv/helix/users', {
        headers: {
          'Client-ID': process.env.REACT_APP_TWITCH_CLIENT_ID,
          Authorization: `Bearer ${oauth.access_token}`
        }
      })
      .then(r => r.json())
      .then(userInfo => {
        //console.log(userInfo); //login [aka lowercase username?], display_name, profile_image_url, description
        localStorage.setItem('__username', userInfo.data[0].login);
        return fetch(`https://api.twitch.tv/helix/moderation/moderators?broadcaster_id=${userInfo.data[0].id}`, {
          headers: {
            'Client-ID': process.env.REACT_APP_TWITCH_CLIENT_ID,
            Authorization: `Bearer ${this.state.access_token}`
          }
        })
        .then(r => r.json())
        .then(modInfo => {
          const modList = modInfo.data.map((modObj) => modObj.user_name.toLowerCase())
          this.setState((state) => {
            return {
              ...state,
              username: userInfo.data[0].login,
              modList
            };
          })
        })
      })
    })
  }

  render() {
    return (
      <div>
        {this.props.location.pathname === "/gamelist" && <JackboxGameList /> }
        {this.state.failed_login && this.props.location.pathname !== "/gamelist"
          ? <Redirect to="/login" />
          : this.state.username && <MainScreen
                                      channel={this.state.username}
                                      modList={this.state.modList}
                                      access_token={this.state.access_token}
                                      onLogout={this.logOut}
                                    />}
      </div>
    )
  }
}


export default withRouter(AuthenticatedApp);
