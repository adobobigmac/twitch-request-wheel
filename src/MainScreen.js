import { Component } from 'react';
import WheelComponent from 'react-wheel-of-prizes'
import GameRequest from './GameRequest'
import MessageHandler from './MessageHandler';
import Sidebar from './Sidebar'
import PlayerSelect from './PlayerSelect';
import ChatActivity, { ActivityStatus } from './ChatActivity';

import './MainScreen.css';
const randomColor = require('randomcolor');


export default class MainScreen extends Component {
    constructor(props){
        super(props);
        this.chatActivity = new ChatActivity(this.props.channel)
        this.state = {
            messages: {},
            colors: randomColor({count: 99, luminosity: 'light', hue: 'blue'}),
            counter: 0,
            history: [],
            nextGameIdx: 0,
            showPlayerSelect: false
        };

        this.playerSelector = null;
        this.messageHandler = null;

        this.changeNextGameIdx = this.changeNextGameIdx.bind(this);
        this.moveNextGameFwd = this.moveNextGameFwd.bind(this);
        this.moveNextGameBack = this.moveNextGameBack.bind(this);
        this.addGameRequest = this.addGameRequest.bind(this);
        this.toggleLock = this.toggleLock.bind(this);
        this.setNextGame = this.setNextGame.bind(this);
        this.addGameToQueue = this.addGameToQueue.bind(this);
        this.onWheelSpun = this.onWheelSpun.bind(this);
        this.removeGame = this.removeGame.bind(this);
        this.onMessage = this.onMessage.bind(this);
        this.togglePlayerSelect = this.togglePlayerSelect.bind(this);
        this.routePlayRequest = this.routePlayRequest.bind(this);
        this.routeLeaveRequest = this.routeLeaveRequest.bind(this);
        this.routeOpenQueueRequest = this.routeOpenQueueRequest.bind(this);
        this.routeCloseQueueRequest = this.routeCloseQueueRequest.bind(this);
        this.routeClearQueueRequest = this.routeClearQueueRequest.bind(this);
        this.startGame = this.startGame.bind(this);
        this.setMessageHandlerRef = this.setMessageHandlerRef.bind(this);
        this.setPlayerSelectRef = this.setPlayerSelectRef.bind(this);
    }

    changeNextGameIdx = (delta = 1) => {
        if (this.state.nextGameIdx + delta > this.state.history.length) return false;
        if (this.state.nextGameIdx + delta < 0) return false;
        this.setState((state) => {
            return {
                ...this.state,
                nextGameIdx: state.nextGameIdx + delta
            }
        });
        return true;
    }

    moveNextGameFwd = () => {
        return this.changeNextGameIdx();
    }

    moveNextGameBack = () => {
        return this.changeNextGameIdx(-1);
    }

    addGameRequest = (gameObj, user) => {
        this.setState((state) => {
            return {
                ...state,
                messages: {
                    ...this.state.messages,
                    [gameObj.longName]: {
                        ...gameObj,
                        username: user,
                        time: Date.now(),
                        locked: false,
                        chosen: false
                    }
                },
                counter: this.state.counter + 1
            };
        })
    }

    toggleLock = (game) => {
        const stateCopy = {...this.state.messages[game]};
        stateCopy.locked = !stateCopy.locked

        this.setState(() => {
            return {
                ...this.state,
                messages: {
                    ...this.state.messages,
                    [game]: stateCopy
                }
            }
        });
    }

    // @return: the number of games ahead of this one, after successfully inserting in queue
    // (i.e. if it's the very next game, return 0; if there's one ahead, return 1; etc)
    setNextGame = (gameObj) => {
        let idx = this.state.nextGameIdx;

        // insert next game at next up position by default, but
        //      *after* any other manually inserted games
        while (idx < this.state.history.length && this.state.history[idx]?.override) {
            idx++;
        }

        this.setState((state) => {
            return {
                ...state,
                history: [
                    ...state.history.slice(0, Math.max(0, idx)),
                    {
                        ...gameObj,
                        override: true
                    },
                    ...state.history.slice(idx)
                ]
            }
        });

        return idx - this.state.nextGameIdx;
    }

    addGameToQueue = (gameObj) => {
        // update history + game card highlight color
        this.setState((state) => {
            return {
                ...state,
                history: [
                    ...this.state.history,
                    {
                        ...gameObj,
                        override: false
                    }
                ],
                messages: {
                    ...state.messages,
                    [gameObj.longName]: {
                        ...state.messages[gameObj.longName],
                        chosen: true
                    }
                }
            }
        })
    }

    onWheelSpun = (gameLongName) => {
        const gameRequestObj = this.state.messages?.[gameLongName];
        if (!gameRequestObj) return;

        // send confirmation message in chat
        const requester = gameRequestObj.username;

        this.addGameToQueue(gameRequestObj);

        // remove card unless it's locked
        if (!this.state.messages[gameLongName].locked) {
            setTimeout(() => {
                this.removeGame(gameLongName);
            }, 2500);
        }

        return this.chatActivity.getStatusPromise(requester).then((status) => {
            let msg = "";
            switch(status) {
                case ActivityStatus.DISCONNECTED:
                    msg = `/me ${gameRequestObj.name} just won the spin, but it doesn't seem like @${requester} is still around. Hope someone else wants to play!`
                    break;

                case ActivityStatus.ACTIVE:
                    msg = `/me @${requester}, good news - ${gameRequestObj.name} just won the spin!`;
                    break;

                case ActivityStatus.IDLE:
                default:
                    msg = `/me @${requester}, good news - ${gameRequestObj.name} just won the spin! (I hope you're still around!)`;
            }
            return this.messageHandler.sendMessage(msg);
        })

    }

    removeGame = (gameLongName) => {
        const newMessageObj = {...this.state.messages};
        delete newMessageObj[gameLongName];
        this.setState((state) => {
            return {
                ...state,
                messages: newMessageObj,
                counter: this.state.counter + 1
            };
        })
    }

    onMessage = (message, user, metadata) => {
        this.chatActivity.updateLastMessageTime(user);
    }

    togglePlayerSelect = () => {
        this.setState((state) => {
            return {
                ...state,
                showPlayerSelect: !state.showPlayerSelect
            }
        })
    }

    routePlayRequest = (user, {sendConfirmationMsg = true, isPrioritySeat = false}) => {
        const msg = this.state.showPlayerSelect
            ? this.playerSelector?.handleNewPlayerRequest(user, {isPrioritySeat})
            : 'sign-ups are currently closed; try again after this game wraps up!'

        if (sendConfirmationMsg) {
            this.messageHandler?.sendMessage(`/me @${user}, ${msg}`);
        }
    }

    routeLeaveRequest = (user) => {
        this.playerSelector?.removeUser(user);
    }

    routeOpenQueueRequest = () => {
        this.setState((state) => {
            return {
                ...state,
                showPlayerSelect: true
            }
        })
        this.playerSelector?.openQueue();
    }

    routeCloseQueueRequest = () => {
        this.playerSelector?.closeQueue();
    }

    routeClearQueueRequest = () => {
        this.playerSelector?.clearQueue();
    }

    startGame = () => {
        // I know this is a big ol' React sin, but I can't for the life of me
        //   figure out why this.togglePlayerSelect() isn't working... sooo...
        if (this.state.showPlayerSelect) {
            this.state.showPlayerSelect = false;
            this.moveNextGameFwd();
            return true;
        }
        return false;
    }

    setMessageHandlerRef = (ps) => {
        this.messageHandler = ps;
    };

    setPlayerSelectRef = (mh) => {
        this.playerSelector = mh;
    };

    render() {
        const gameRequestArray = Object.keys(this.state.messages);

        let logOutBtn;
        if (typeof this.props.onLogout === 'function') {
            logOutBtn = (
                <button className="logout" onClick={this.props.onLogout}>Logout &#10151;</button>
            );
        }

        let mainClassName = this.state.showPlayerSelect ? 'player-select' : 'game-select';

        let subheading = this.state.showPlayerSelect ? (
            <span className="subheading-player">
                Type <b>!new</b> in {this.props.channel}'s chat if you want to join the next game
            </span>
        ) : (
            <span className="subheading-game">
                Type e.g. <b>"!request Blather Round"</b> in {this.props.channel}'s chat to add
            </span>
        );

        let innerContent;
        let rightColumn;

        if (this.state.showPlayerSelect) {
            innerContent = (
                <PlayerSelect
                    game={this.state.history?.[this.state.nextGameIdx]}
                    startGame={this.startGame}
                    ref={this.setPlayerSelectRef}
                />
            );
        } else {
            innerContent = gameRequestArray.map((gameName, i) =>
                <GameRequest
                    key={i}
                    gameName={gameName}
                    metadata={this.state.messages[gameName]}
                    onDelete={this.removeGame}
                    toggleLock={this.toggleLock.bind(gameName)}
                    getActivity={this.chatActivity.getStatusPromise}
                />
            );
            rightColumn = (
                <div className="right-column" width="50px">
                    <div className="wheel-wrapper">
                         <WheelComponent
                            key={this.state.counter}
                            segments={gameRequestArray}
                            segColors={this.state.colors}
                            onFinished={this.onWheelSpun}
                            isOnlyOnce={false}
                            size={250}
                            upDuration={100}
                            downDuration={1000}
                            primaryColor={"white"}
                            contrastColor={"black"}
                            fontFamily={"Arial"}
                        />
                    </div>
                </div>
            );
        }


        return (
            <div id="main-screen" className={mainClassName}>
                <MessageHandler
                    addGameRequest={this.addGameRequest}
                    setNextGame={this.setNextGame}
                    changeNextGameIdx={this.changeNextGameIdx}
                    startGame={this.startGame}
                    messages={this.state.messages}
                    channel={this.props.channel}
                    modList={this.props.modList}
                    access_token={this.props.access_token}
                    onMessage={this.onMessage}
                    onDelete={this.removeGame}
                    upcomingGames={this.state.history.slice(this.state.nextGameIdx)}
                    caniplayHandler={this.routePlayRequest}
                    playerExitHandler={this.routeLeaveRequest}
                    openQueueHandler={this.routeOpenQueueRequest}
                    closeQueueHandler={this.routeCloseQueueRequest}
                    clearQueueHandler={this.routeClearQueueRequest}
                    ref={this.setMessageHandlerRef}
                />
                <div className="left-column">
                    <h2>{this.state.showPlayerSelect ? 'Seat Requests' : 'Game Requests'}</h2>
                    <h4>{subheading}</h4>
                    <div className="left-column-body">
                        <Sidebar
                            history={this.state.history}
                            nextGameIdx={this.state.nextGameIdx}
                            changeNextGameIdx={this.changeNextGameIdx}
                            moveNextGameFwd={this.moveNextGameFwd}
                            moveNextGameBack={this.moveNextGameBack}
                            togglePlayerSelect={this.togglePlayerSelect}
                            requestMode={this.state.showPlayerSelect ? 'seat' : 'game'}
                        />
                        <div className="left-column-inner-body">
                            {innerContent}
                        </div>
                    </div>
                </div>
                {rightColumn}
                {logOutBtn}
            </div>
        )
    }
}
