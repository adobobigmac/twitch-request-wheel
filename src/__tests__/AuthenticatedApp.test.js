import {AuthenticatedApp} from '../AuthenticatedApp';
import {createRenderer} from 'react-test-renderer/shallow';
import MainScreen from '../MainScreen';
import React from 'react';
import {Redirect} from "react-router-dom";
import {render} from '@testing-library/react';

jest.mock('../MainScreen');
jest.mock('react-router-dom', () => {
    const reactRouterDom = jest.requireActual('react-router-dom');
    return {
        ...reactRouterDom,
        Redirect: () => () => (<div>React Tooltip Mock</div>)
    };
});

describe('AuthenticatedApp', () => {
    // TODO: fix the following error from AuthenticatedApp (when importing the default class):
    // Error: Invariant failed: You should not use <withRouter(AuthenticatedApp) /> outside a <Router>
    let props = {
    	history: {
    		length: 3,
    		action: 'POP',
    		location: {
    			pathname: '/',
    			search: '?code=foobar&scope=chat:read chat:edit moderation:read',
    			hash: ""
    		}
    	},
    	location: {
    		pathname: '/',
    		search: '?code=foobar&scope=chat:read chat:edit moderation:read',
    		hash: ""
    	},
    	match: {
    		path: '/',
    		url: '/',
    		isExact: true,
    		params: {}
    	}
    };
    test('render page', async () => {
        let {container} = render(<AuthenticatedApp {...props} />);
        expect(container).toMatchSnapshot();
        expect(container.querySelector('div > div')).toBeDefined();
    });
    test('render with MainScreen', () => {
        const shallowRenderer = createRenderer();
        shallowRenderer.render(<AuthenticatedApp {...props} />);
        let instance = shallowRenderer.getMountedInstance();
        instance.setState({
            access_token: 'yadayadayada',
            failed_login: false,
            modList: [],
            username: 'sirgoosewell'
        });
        let component = shallowRenderer.getRenderOutput();
        expect(component.props.children.type).toBe(MainScreen);
        expect(component).toMatchSnapshot();
        shallowRenderer.unmount();
    });
    test('render with Redirect', () => {
        const shallowRenderer = createRenderer();
        shallowRenderer.render(<AuthenticatedApp {...props} />);
        let instance = shallowRenderer.getMountedInstance();
        instance.setState({
            access_token: null,
            failed_login: true,
            username: null
        });
        let component = shallowRenderer.getRenderOutput();
        expect(component.type).toBe(Redirect);
        expect(component).toMatchSnapshot();
        shallowRenderer.unmount();
    });
});
