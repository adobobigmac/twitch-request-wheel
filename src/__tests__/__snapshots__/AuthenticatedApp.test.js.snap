// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticatedApp render page 1`] = `
<div>
  <div
    id="authenticated-app"
  />
</div>
`;

exports[`AuthenticatedApp render with MainScreen 1`] = `
<div
  id="authenticated-app"
>
  <MainScreen
    access_token="yadayadayada"
    channel="sirgoosewell"
    modList={Array []}
    onLogout={[Function]}
  />
</div>
`;

exports[`AuthenticatedApp render with Redirect 1`] = `
<Redirect
  to="/login"
/>
`;
